syspagen/syspagen.c

```diff
@@ -0,0 +1,784 @@
+{
+	FILE *file;
+	size_t len = 0;
+	char *line = NULL;
+	int sz, argc, i, ret;
+	char *argv[SIZE_CMD_ARGV];
+	char argline[SIZE_CMD_ARG_LINE];
+
+	file = fopen(fname, "r");
+	if (file == NULL) {
+		fprintf(stderr, "Cannot open file %s\n", fname);
+		return -1;
+	}
+
+	while ((sz = getline(&line, &len, file)) != -1) {
+		argc = sysgen_parseArgLine(line, argline, SIZE_CMD_ARG_LINE, argv, SIZE_CMD_ARGV);
+		/* skip empty lines */
+		if (argc == 0) {
+			continue;
+		}
+		/* either end of script or no script */
+		else if (argc == -1) {
+			free(line);
+			fclose(file);
+			return 0;
+		}
+		/* error */
+		else if (argc < 0) {
+			free(line);
+			fclose(file);
+			return argc;
+		}
+
+		/* Find command and launch associated function */
+		for (i = 0; i < sizeof(cmds) / sizeof(cmd_t); ++i) {
+			if (strcmp(argv[0], cmds[i].name) != 0)
+				continue;
+
+			if ((ret = cmds[i].run(argc, argv)) < 0) {
+				fprintf(stderr, "Failed %s\n", argv[0]);
+				free(line);
+				fclose(file);
+				return ret;
+			}
+
+			break;
+		}
+	}
+
+	free(line);
+	fclose(file);
+
+	return 0;
+}
+
+
+static int sysgen_addSyspage2Img(const char *imgName)
+{
+	int res;
+	size_t sz;
+	FILE *img;
+
+	img = fopen(imgName, "rb+");
+	if (img == NULL)
+		return -1;
+
+	res = fseek(img, sysgen_common.offs, SEEK_SET);
+	if (res < 0) {
+		fclose(img);
+		return -1;
+	}
+
+	sz = fwrite(sysgen_common.buff, sizeof(uint8_t), sysgen_common.syspage->size, img);
+	if (sz < sysgen_common.syspage->size) {
+		fprintf(stderr, "Cannot write binary syspage into %s\n", imgName);
+		fclose(img);
+		return -1;
+	}
+
+	fclose(img);
+
+	return 0;
+}
+
+
+static void sysgen_cleanup(void)
+{
+	struct phfs_alias_t *alias;
+
+	while (!LIST_EMPTY(&sysgen_common.aliases)) {
+		alias = LIST_FIRST(&sysgen_common.aliases);
+		LIST_REMOVE(alias, ptrs);
+		free(alias);
+	}
+
+	free(sysgen_common.buff);
+}
+
+
+static void sysgen_dump()
+{
+	sysptr_t ptr;
+	const syspage_prog_t *prog;
+
+	printf("\n\tSyspage:\n");
+	printf("\tImage size: 0x%08x\n", sysgen_common.syspage->hs.imgsz);
+	printf("\tSyspage size: 0x%08x\n", sysgen_common.syspage->size);
+	printf("\tKernel physical address: 0x%08x\n", sysgen_common.syspage->pkernel);
+	printf("\tConsole: 0x%02x\n", sysgen_common.syspage->console);
+	printf("\tPrograms:\n");
+	if (sysgen_common.syspage->progs != 0) {
+		ptr = sysgen_common.syspage->progs;
+		do {
+			prog = CAST_SYSPTR(syspage_prog_t *, ptr);
+			printf("\t\t%s\n", CAST_SYSPTR(char *, prog->argv));
+		} while ((ptr = prog->next) != sysgen_common.syspage->progs);
+	}
+	else {
+		printf("\t\tnot defined\n");
+	}
+}
+
+
+static void sysgen_help(const char *prog)
+{
+	printf("Usage: %s [options] to create binary syspage file or add syspage to image\n", prog);
+	printf("\t-s <pimg> <offs> <sz>   - syspage properties\n");
+	printf("\t    pimg - beginning physical address of the target image\n");
+	printf("\t    offs - syspage's offset in the target image\n");
+	printf("\t    sz   - max syspage's size\n");
```

Comment:
Maybe this will be better
```suggestion
	printf("\t-s <pimg:offs:sz>   - syspage properties\n");
	printf("\t    pimg - beginning physical address of the target image\n");
	printf("\t    offs - syspage's offset in the target image\n");
	printf("\t    sz   - max syspage's size\n");
```

Reply:
Thanks for suggestion, implemented.